File: Devil_Java/src/main/java/week03_greedy/sohi/BOJ1744.java
package week03_greedy.sohi;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BOJ1744 {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        // 음수, 양수 나눠서 담기
        List<Integer> positive = new ArrayList<>();
        List<Integer> negative = new ArrayList<>();
        int zero = 0, one = 0, result = 0;

        for (int i = 0; i < N; i++) {
            int num = Integer.parseInt(br.readLine());
            if (num > 1) positive.add(num);
            else if (num == 1) one++;
            else if (num == 0) zero++;
            else negative.add(num);
        }

        result += one; // 1 걍 더하고

        positive.sort(Collections.reverseOrder());
        Collections.sort(negative);

        for (int i = 0; i < positive.size(); i = i + 2) {
            if (positive.size() > i + 1) {
                result += positive.get(i) * positive.get(i + 1);
            } else result += positive.get(i);

        }

        for (int i = 0; i < negative.size(); i += 2) {
            if (i + 1 < negative.size()) {
                result += negative.get(i) * negative.get(i + 1);
            } else if (zero == 0) result += negative.get(i);

        }

        System.out.println(result);

    }

}

File: Devil_Python/week03_recursion/gold/BOJ1744.py
# 합이 최대가 되게 두 수 곱해서 더하기
# 음수, 0, 1, 양수
# 음수는 음수끼리 양수는 양수끼리 / 음수 0하고 곱하기 1은 걍 더하기

N = int(input().strip())
nums = [int(input().strip()) for _ in range(N)]


#nums.sort()

# 음수, 양수, 따로 정렬?

negative = []
zero = 0
one = 0
positive = []
result = 0

for n in nums:
    if  n > 1:
        positive.append(n)
    elif n < 0:
        negative.append(n)
    elif n == 1:
        one += 1
    elif n == 0:
        zero += 1

positive.sort(reverse=True)
negative.sort()

# 두 개씩 묶어서 곱하기 음수, 양수
for i in range(0, len(negative), 2):
    if i + 1 < len(negative):
        result += negative[i] * negative[i+1]
    elif zero == 0:
        result += negative[i]

for i in range(0, len(positive), 2):
    if i + 1 < len(positive):
        result += positive[i] * positive[i+1]
    else:
        result += positive[i]

result += one

print(result)




